cmd/lakectl/cmd/fs.go

```diff
@@ -409,6 +413,164 @@ var fsRmCmd = &cobra.Command{
+
+		if parallel < 1 {
+			DieFmt("Invalid value for parallel (%d), minimum is 1.\n", parallel)
+		}
+
+		// optional destination directory
+		var dest string
+		if len(args) > 1 {
+			dest = args[1]
+			// verify destination path exists
+			if len(dest) > 0 {
+				pathExists(dest)
+			}
+		}
+
+		// list the files
+		client := getClient()
+		downloadCh := make(chan string)
+		sourcePath := api.StringValue(pathURI.Path)
+
+		// recursive assume source is directory
+		if recursive && len(sourcePath) > 0 && !strings.HasSuffix(sourcePath, "/") {
+			sourcePath += "/"
+		}
+
+		// prefix to remove from destination
+		prefix := filepath.Dir(sourcePath)
+		if prefix != "" {
+			prefix += "/"
+		}
+		ctx := cmd.Context()
+		// list objects to download
+		go func() {
+			defer close(downloadCh)
+			if recursive {
+				listRecursiveHelper(ctx, client, pathURI.Repository, pathURI.Ref, sourcePath, downloadCh)
+			} else {
+				downloadCh <- api.StringValue(pathURI.Path)
+			}
+		}()
+
+		// download in parallel
+		var (
+			wg         sync.WaitGroup
+			errCounter int64
+		)
+		wg.Add(parallel)
+		for i := 0; i < parallel; i++ {
+			go func() {
+				defer wg.Done()
+				for downloadPath := range downloadCh {
+					src := uri.URI{
+						Repository: pathURI.Repository,
+						Ref:        pathURI.Ref,
+						Path:       &downloadPath,
+					}
+					// destination is without the source URI
+					dst := filepath.Join(dest, strings.TrimPrefix(downloadPath, prefix))
+					err := downloadHelper(ctx, client, direct, src, dst)
+					if err == nil {
+						fmt.Printf("Download: %s to %s\n", src.String(), dst)
+					} else {
+						_, _ = fmt.Fprintf(os.Stderr, "Download failed: %s to %s - %s\n", src.String(), dst, err)
+						atomic.AddInt64(&errCounter, 1)
+					}
+				}
+			}()
+		}
+
+		// wait for download to complete
+		wg.Wait()
+		// exit with the right status code
+		if atomic.LoadInt64(&errCounter) > 0 {
+			defer os.Exit(1)
+		}
+	},
+}
+
+func listRecursiveHelper(ctx context.Context, client *api.ClientWithResponses, repo, ref, prefix string, ch chan string) {
+	pfx := api.PaginationPrefix(prefix)
+	var from string
+	for {
+		params := &api.ListObjectsParams{
+			Prefix: &pfx,
+			After:  api.PaginationAfterPtr(from),
+		}
+		resp, err := client.ListObjectsWithResponse(ctx, repo, ref, params)
+		DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusOK)
+		for _, p := range resp.JSON200.Results {
+			ch <- p.Path
+		}
+		pagination := resp.JSON200.Pagination
+		if !pagination.HasMore {
+			break
+		}
+		from = pagination.NextOffset
+	}
+}
+
+func downloadHelper(ctx context.Context, client *api.ClientWithResponses, direct bool, src uri.URI, dst string) error {
+	body, err := getObjectHelper(ctx, client, direct, src)
+	if err != nil {
+		return err
+	}
+	defer func(body io.ReadCloser) {
+		_ = body.Close()
+	}(body)
+
+	// create destination dir if needed
+	dir := filepath.Dir(dst)
+	_ = os.MkdirAll(dir, os.ModePerm)
```

Comment:
We allow an object and a prefix of the same name. I think it might be problematic, due to the behavior of [`os.MkDirAll`](https://github.com/golang/go/blob/9e7c5030d064395e24a25dd1a81722f5583defcf/src/os/path.go#L25), as the object will be created as a local file, failing a firther attempt to create it as a directory. 
```
$ ./lakectl fs ls lakefs://myrepo/main/external/
object          2022-10-21 12:09:30 -0500 CDT    3.2 kB          internal
common_prefix                                                    internal/

$ ./lakectl fs ls lakefs://myrepo/main/external/ --recursive
object          2022-10-21 12:09:30 -0500 CDT    3.2 kB          external/internal
object          2022-10-21 12:09:36 -0500 CDT    3.2 kB          external/internal/obj1
```

Reply:
Thinking of it thourhg, this is actually a problem with the different way prefixes are treated in `lakefs` (and probably object stores ?) vs the way OS handles dirs. Not sure what the solution should be

Reply:
- about the fail cascade in case of write that fail, it can occur for example when we try to download to a place we can't write. all the lines will report an error that we failed.
- about the behavior of 'ls' and 'download' regarding local files and prefix - we captured aws S3 behavior.